modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/TypedInput.es.js

```diff
@@ -1,37 +1,178 @@
+		onChange: propTypes.func.isRequired,
+		options: propTypes.array,
+		type: propTypes.string.isRequired,
+		value: propTypes.oneOfType(
+			[
+				propTypes.string,
+				propTypes.number
+			]
+		)
+	};
+
+	static defaultProps = {
+		options: []
+	};
+
+	constructor(props) {
+		super(props);
+
+		const {value} = props;
+
+		this.state = {
+			decimal: value
+		};
+	}
+
+	_getInputRenderer(type) {
+		const renderers = {
+			[PROPERTY_TYPES.BOOLEAN]: this._renderBooleanType,
+			[PROPERTY_TYPES.DATE]: this._renderDateType,
+			[PROPERTY_TYPES.DOUBLE]: this._renderDecimalType,
+			[PROPERTY_TYPES.INTEGER]: this._renderIntegerType,
+			[PROPERTY_TYPES.STRING]: this._renderStringType
+		};
+
+		return renderers[type] || renderers[PROPERTY_TYPES.STRING];
+	}
+
 	_handleChange = event => {
 		const {onChange} = this.props;
 
 		const value = event.target.value;
 
-		if (onChange) {
-			onChange(value);
+		onChange(value);
+	}
+
+	_handleDateChange = event => {
+		const {onChange} = this.props;
+
+		const value = event.target.value ||
+			dateFns.format(new Date(), INPUT_DATE_FORMAT);
+
+		const iSOStringValue = dateFns
+			.parse(value, INPUT_DATE_FORMAT)
+			.toISOString();
+
+		onChange(iSOStringValue, PROPERTY_TYPES.DATE);
+	}
+
+	_handleDecimalBlur = event => {
+		const {onChange} = this.props;
+
+		const value = Number.parseFloat(event.target.value).toFixed(2);
+
+		this.setState(
+			{
+				decimal: value
+			},
+			onChange(value)
+		);
+	}
+
+	_handleDecimalChange = event => {
+		const value = event.target.value;
+
+		this.setState(
+			{
+				decimal: value
+			}
+		);
+	}
+
+	_handleIntegerChange = event => {
+		const {onChange} = this.props;
+		const value = parseInt(event.target.value, 10);
+
+		if (!isNaN(value)) {
+			onChange(value.toString());
 		}
 	}
 
-	render() {
-		const {options, value} = this.props;
+	_renderBooleanType = () => {
+		const {value} = this.props;
+
+		return (
+			<ClaySelect
+				className="criterion-input form-control"
+				data-testid="options-boolean"
+				onChange={this._handleChange}
+				options={BOOLEAN_OPTIONS}
+				selected={value}
+			/>);
+	}
 
-		const classnames = getCN(
-			'criterion-input',
-			'form-control'
+	_renderDateType = () => {
+		const {value} = this.props;
+
+		const date = new Date(value);
+
+		const domStringDate = dateFns.format(date, INPUT_DATE_FORMAT);
+
+		return (
+			<div className="criterion-input date-input">
+				<input
+					className="form-control"
+					data-testid="date-input"
+					onChange={this._handleDateChange}
+					type="date"
+					value={domStringDate}
+				/>
+			</div>
 		);
+	}
+
+	_renderDecimalType = () => {
+		const {decimal} = this.state;
+
+		return (
+			<input
+				className="criterion-input form-control"
+				data-testid="decimal-number"
+				onBlur={this._handleDecimalBlur}
+				onChange={this._handleDecimalChange}
+				step="0.01"
+				type="number"
+				value={decimal}
+			/>
+		);
+	}
+
+	_renderIntegerType = () => {
+		const {value} = this.props;
```

Comment:
no need to destructure if its only used once and no other values are destructured from props.